ante/fee.go

```diff
@@ -0,0 +1,337 @@
+
+// DeductFeeDecorator deducts fees from the first signer of the tx
+// If the first signer does not have the funds to pay for the fees, return with InsufficientFunds error
+// Call next AnteHandler if fees successfully deducted
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+type DeductFeeDecorator struct {
+	accountKeeper  ante.AccountKeeper
+	bankKeeper     BankKeeper
+	feegrantKeeper ante.FeegrantKeeper
+	txFeeChecker   TxFeeChecker
+}
+
+func NewDeductFeeDecorator(ak ante.AccountKeeper, bk BankKeeper, fk ante.FeegrantKeeper, tfc TxFeeChecker) DeductFeeDecorator {
+	if tfc == nil {
+		tfc = checkTxFeeWithCustomFixedFee
+	}
+
+	return DeductFeeDecorator{
+		accountKeeper:  ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+		txFeeChecker:   tfc,
+	}
+}
+
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if !simulate && ctx.BlockHeight() > 0 && feeTx.GetGas() == 0 {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidGasLimit, "must provide positive gas")
+	}
+
+	var (
+		priority    int64
+		isCustomFee bool
+		err         error
+	)
+
+	fee := feeTx.GetFee()
+	if !simulate {
+		fee, priority, isCustomFee, err = dfd.txFeeChecker(ctx, tx)
+		if err != nil {
+			return ctx, err
+		}
+	}
+
+	if err := dfd.checkDeductFeeWithFixedFee(ctx, tx, isCustomFee, fee, simulate); err != nil {
+		return ctx, err
+	}
+
+	newCtx := ctx.WithPriority(priority)
+
+	return next(newCtx, tx, simulate)
+}
+
+func (dfd DeductFeeDecorator) checkDeductFeeWithFixedFee(ctx sdk.Context, sdkTx sdk.Tx, isCustomFee bool, fee sdk.Coins, simulate bool) error {
+	feeTx, ok := sdkTx.(sdk.FeeTx)
+	if !ok {
+		return sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if !isCustomFee {
+		return dfd.checkDeductFee(ctx, feeTx, fee)
+	}
+
+	if simulate {
+		return sdkerrors.Wrap(sdkerrors.ErrNotSupported, "simulation of fees is not supported for network specific transactions")
+	}
+
+	if addr := dfd.accountKeeper.GetModuleAddress(cheqdtypes.ModuleName); addr == nil {
+		return fmt.Errorf("cheqd fee collector module account (%s) has not been set", cheqdtypes.ModuleName)
+	}
+
+	feePayer := feeTx.FeePayer()
+	feeGranter := feeTx.FeeGranter()
+	deductFeesFrom := feePayer
+
+	// if feegranter set deduct fee from feegranter account.
+	// this works with only when feegrant enabled.
+	if feeGranter != nil {
+		if dfd.feegrantKeeper == nil {
+			return sdkerrors.ErrInvalidRequest.Wrap("fee grants are not enabled")
+		} else if !feeGranter.Equals(feePayer) {
+			err := dfd.feegrantKeeper.UseGrantedFees(ctx, feeGranter, feePayer, fee, sdkTx.GetMsgs())
+			if err != nil {
+				return sdkerrors.Wrapf(err, "%s does not not allow to pay fees for %s", feeGranter, feePayer)
+			}
+		}
+
+		deductFeesFrom = feeGranter
+	}
```

Comment:
Do we currently or in the future plan on using `feegrant` module?

Reply:
Embedded by default in fee-checking closures, for upstream compatibility.
As for the module, `feegrant` enabled by default, so easier to keep than to ditch in terms of effort needed.